bob-common/src/mapper.rs

```diff
@@ -130,26 +168,30 @@ impl Virtual {
         let target_nodes = self.get_target_nodes_for_key(key);
         trace!("extract indexes of target nodes");
         trace!("nodes available: {}", self.nodes.len());
-        let offset = self.support_nodes_offset.fetch_add(1, Ordering::Relaxed);
-        let mut support_nodes = Vec::with_capacity(count);
-        for (id, node) in self.nodes.iter().skip(offset % self.nodes.len()) {
-            if target_nodes.iter().all(|i| i.index() != *id) {
+        let mut support_nodes: Vec<&Node> = Vec::with_capacity(count);
+
+        let starting_index =
+            self.find_actual_offset(self.support_nodes_offset.fetch_add(1, Ordering::Relaxed));
+
+        let len = self.nodes.len();
+        for i in 0..len {
+            let node = &self.nodes[(i + starting_index) % len];
+            let conn = if i + starting_index < len {
+                node.connection_available()
+            } else {
+                true
+            };
+            if conn
+                && target_nodes.iter().all(|n| n.index() != node.index())
+                && (i + starting_index < len
+                    || support_nodes.iter().all(|&n| n.index() != node.index()))
+            {
                 support_nodes.push(node);
                 if support_nodes.len() >= count {
                     break;
                 }
             }
         }
```

Comment:
After this cycle ends, you should check, whether the `support_nodes` list is full. If it is not, then you should add also a disconnected nodes